torch/_functorch/_activation_checkpointing/knapsack.py

```diff
@@ -0,0 +1,123 @@
+from typing import List, Tuple
+
+import torch
+
+
+def greedy_knapsack(
+    memory: List[float], runtimes: List[float], max_memory: float
+) -> Tuple[float, List[int], List[int]]:
+    n = len(runtimes)
+    items = list(range(n))
+
+    # Sort items based on the ratio of runtime to memory in descending order
+    items = sorted(items, key=lambda i: runtimes[i] / memory[i], reverse=True)
+
+    total_memory = 0.0
+    total_runtime = 0.0
+    items_to_save = []
+    items_to_allow_recomputing = []
+
+    for i in items:
+        if total_memory + memory[i] <= max_memory:
+            total_memory += memory[i]
+            total_runtime += runtimes[i]
+            items_to_save.append(i)
+        else:
+            items_to_allow_recomputing.append(i)
+    return total_runtime, items_to_save, items_to_allow_recomputing
+
+
+def ilp_knapsack(
+    memory: List[float], runtimes: List[float], max_memory: float
+) -> Tuple[float, List[int], List[int]]:
+    import numpy as np
+
+    try:
+        from scipy.optimize import Bounds, LinearConstraint, milp
+    except ImportError:
+        raise RuntimeError(
+            "To use the ILP for memory budget checkpointing you need to install scipy"
+        ) from None
```

Comment:
Ugh, why doesn't this just raise from the ImportError. Really unclear why erasing all the useful import error data from the trace is useful.

Reply:
This PR is just moving logic from the partitioners file to a separate file. 

Maybe @Chillee can answer your question. Logic was written as part of https://github.com/pytorch/pytorch/pull/126320?fbclid=IwZXh0bgNhZW0CMTEAAR19wrq3QY7vibee0B_mabKSFaxr54DE6EXml4sxvjEMDTcbPE9UsoL2aDE_aem_n0yhwybp_d4eiz4LAiv7JQ